### PR TITLE
236 modified cosine score

### LINKED ISSUE #26: Missing modified score
So far we implemented the cosine score. The other main score we work with in the iomega project is called "modified cosine score" and in addition takes into account mass-shifts of peaks due to different precursor-m/z.

### PR DESCRIPTION
**Pull request by jspaaks**
Originally here https://github.com/matchms/matchms-backup/pull/237

 -  Added ``ModifiedCosine`` class to similarities.
 -  Added unit test.
 -  Added ``numba`` to environment.

### CODE REVIEW

File: matchms/similarity/ModifiedCosine.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import numba
```
Comment (sverhoeven) [inline matchms/similarity/ModifiedCosine.py]:
```suggestion
from typing import Tuple
import numba
```

File: matchms/similarity/ModifiedCosine.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import numpy
+from matchms.typing import SpectrumType
+
+
+class ModifiedCosine:
+    """Calculate modified cosine score between spectra.
```
Comment (sverhoeven) [inline matchms/similarity/ModifiedCosine.py]: Can you explain what the modified cosine similarity is or add a link where it is described.

File: matchms/similarity/ModifiedCosine.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    Args:
+    ----
+    tolerance: float
+        Peaks will be considered a match when <= tolerance appart. Default is 0.1.
+    """
+    def __init__(self, tolerance=0.1):
+        self.tolerance = tolerance
+
+    def __call__(self, spectrum1: SpectrumType, spectrum2: SpectrumType) -> float:
```
Comment (sverhoeven) [inline matchms/similarity/ModifiedCosine.py]: It returns score and nr matches
```suggestion
    def __call__(self, spectrum1: SpectrumType, spectrum2: SpectrumType) -> Tuple[float, int]:
```

File: matchms/similarity/ModifiedCosine.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        def get_matching_pairs():
+            """Find all pairs of peaks that match within the given tolerance."""
+            zero_pairs = find_pairs_numba(spec1, spec2, self.tolerance, shift=0.0)
+            if spectrum1.get("precursor_mz") and spectrum2.get("precursor_mz"):
+                mass_shift = spectrum1.get("precursor_mz") - spectrum2.get("precursor_mz")
+                nonzero_pairs = find_pairs_numba(spec1, spec2, self.tolerance, shift=mass_shift)
+            else:
+                print("Precursor_mz missing. Calculate cosine score instead.",
+                      "Consider applying 'add_precursor_mz' filter first.")
+                nonzero_pairs = []
```
Comment (sverhoeven) [inline matchms/similarity/ModifiedCosine.py]: A print statement can be missed, please change to assertion so code raises exception
```suggestion
            message = "Precursor_mz missing. Apply 'add_precursor_mz' filter first."
            assert spectrum1.get("precursor_mz") and spectrum2.get("precursor_mz"), message
            mass_shift = spectrum1.get("precursor_mz") - spectrum2.get("precursor_mz")
            nonzero_pairs = find_pairs_numba(spec1, spec2, self.tolerance, shift=mass_shift)
```

File: matchms/similarity/ModifiedCosine.py (hunk 2 above)
Comment (florian-huber) [reply matchms/similarity/ModifiedCosine.py]: I added a bit more extensive explanation and a references for further details.

File: matchms/similarity/ModifiedCosine.py (hunk 4 above)
Comment (florian-huber) [reply matchms/similarity/ModifiedCosine.py]: That's better, thanks!

File: matchms/similarity/ModifiedCosine.py (hunk 2 above)
Comment (sverhoeven) [reply matchms/similarity/ModifiedCosine.py]: nice

File: matchms/similarity/ModifiedCosine.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        self.tolerance = tolerance
+
+    def __call__(self, spectrum1: SpectrumType, spectrum2: SpectrumType) -> Tuple[float, int]:
+        """Calculate modified cosine score between two spectra.
+        Args:
+        ----
+        spectrum1: SpectrumType
+            Input spectrum 1.
+        spectrum2: SpectrumType
+            Input spectrum 2.
```
Comment (sverhoeven) [inline matchms/similarity/ModifiedCosine.py:34]: Describe return
```suggestion
            Input spectrum 2.

Returns:
--------

Tuple with cosine score and number of matched peaks.
```

**Review by sverhoeven — CHANGES_REQUESTED** [review]
Tested it without numba annotation my test took 48s, with took 28s

**Comment by florian-huber** [comment]
> 
> 
> Tested it without numba annotation my test took 48s, with took 28s

I usually saw more a speed-up of about  5-6x, but that might also depend on the data I assume.

[edit]: I now tested it again with generated spectrums:
```python
n_peaks = 200
n_spectrums = 100

# Generate synthetic data
test_spectrums = []
for i in range(n_spectrums):
    test_spectrum = Spectrum(mz=numpy.linspace(10, 1000, n_peaks),
                             intensities=numpy.random.random(n_peaks),
                             metadata={"precursor_mz": numpy.random.randint(100,110)})
    test_spectrums.append(test_spectrum)

# Run with numba
similarity_measure = ModifiedCosine(tolerance=0.01)

tstart = time.time()
similarity_matrix = Scores(test_spectrums, test_spectrums,
                           similarity_measure).calculate().scores
tend = time.time()
print("execution time:", tend-tstart)

# Run without numba
similarity_measure = ModifiedCosine_no_numba(tolerance=0.01)

tstart = time.time()
similarity_matrix = Scores(test_spectrums, test_spectrums,
                           similarity_measure).calculate().scores
tend = time.time()
print("execution time:", tend-tstart)
```
And I got ``4.13...``s with numba versus ``29.47...``s without numba.

File: matchms/similarity/ModifiedCosine.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        def get_matching_pairs():
+            """Find all pairs of peaks that match within the given tolerance."""
+            zero_pairs = collect_peak_pairs(spec1, spec2, self.tolerance, shift=0.0)
+            message = "Precursor_mz missing. Apply 'add_precursor_mz' filter first."
+            assert spectrum1.get("precursor_mz") and spectrum2.get("precursor_mz"), message
+            mass_shift = spectrum1.get("precursor_mz") - spectrum2.get("precursor_mz")
+            nonzero_pairs = collect_peak_pairs(spec1, spec2, self.tolerance, shift=mass_shift)
+            matching_pairs = zero_pairs + nonzero_pairs
+            matching_pairs = sorted(matching_pairs, key=lambda x: x[2], reverse=True)
+            return matching_pairs
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/ModifiedCosine.py]: That looks cleaner, thanks.

**Comment by sverhoeven** [comment]
My speed test uses the profiling script

```python
import os
from matchms import Scores
from matchms.filtering import add_parent_mass, add_precursor_mz
from matchms.filtering import default_filters
from matchms.filtering import normalize_intensities
from matchms.filtering import require_minimum_number_of_peaks
from matchms.filtering import select_by_mz
from matchms.filtering import select_by_relative_intensity
from matchms.importing import load_from_mgf
from matchms.similarity import ModifiedCosine


def apply_my_filters(s):
    s = default_filters(s)
    s = add_parent_mass(s)
    s = normalize_intensities(s)
    s = select_by_relative_intensity(s, intensity_from=0.5, intensity_to=1.0)
    s = select_by_mz(s, mz_from=0, mz_to=1000)
    s = require_minimum_number_of_peaks(s, n_required=5)
    return s


def main():
    spectrums_file = os.path.join("GNPS-LIBRARY.mgf")

    # apply my filters to the data
    print('Reading')
    spectrums = [apply_my_filters(s) for s in load_from_mgf(spectrums_file)]

    # omit spectrums that didn't qualify for analysis
    spectrums = [s for s in spectrums if s is not None and s.get('precursor_mz')]

    # define similarity function
    scorer = ModifiedCosine()

    # this will be a library grouping analysis, so queries = references = spectrums
    queries = spectrums[:]
    references = spectrums[:]

    print(len(queries))
    print('Scoring')
    similarity_matrix = Scores(references=references,
                               queries=queries,
                               similarity_function=scorer).calculate().scores


main()

```

**Review by sverhoeven — APPROVED** [review]

**Comment by florian-huber** [comment]
Thanks a lot for the review @sverhoeven!

**Pull request merged**